lib/router/src/lib.rs

```diff
@@ -208,43 +247,80 @@ where
     fn not_recognized() -> Self {
         Self::error(error::NotRecognized.into())
     }
-
-    fn no_capacity(capacity: usize) -> Self {
-        Self::error(error::NoCapacity(capacity).into())
-    }
 }
 
-impl<Req, Svc> Future for ResponseFuture<Req, Svc>
+impl<Req, Rec, Mk> Future for ResponseFuture<Req, Rec, Mk>
 where
-    Svc: svc::Service<Req>,
-    Svc::Error: Into<error::Error>,
+    Rec: Recognize<Req>,
+    Mk: Make<Rec::Target>,
+    Mk::Value: svc::Service<Req> + Clone,
+    <Mk::Value as svc::Service<Req>>::Error: Into<error::Error>,
 {
-    type Item = <LoadShed<Svc> as svc::Service<Req>>::Response;
+    type Item = <LoadShed<Mk::Value> as svc::Service<Req>>::Response;
     type Error = error::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         use svc::Service;
 
         loop {
             self.state = match self.state {
-                State::Init(ref mut req, ref mut svc) => {
-                    match svc.poll_ready().map_err(error::Error::from)? {
-                        Async::NotReady => {
-                            unreachable!("load shedding services must always be ready");
-                        }
-                        Async::Ready(()) => {
-                            let req = req.take().expect("polled after ready");
-                            State::Called(svc.call(req))
-                        }
+                State::Inserting(ref mut rr) => {
+                    // Aquire the lock for the router cache
+                    let mut cache = match rr.cache.poll_lock() {
+                        Async::Ready(aquired) => aquired,
+                        Async::NotReady => return Ok(Async::NotReady),
+                    };
+
+                    let request = rr.request.take().expect("polled after ready");
+                    let target = rr.target.take().expect("polled after ready");
+
+                    // If the target is already cached, route the request to
+                    // the service; otherwise, try to insert it
+                    if let Some(service) = cache.get(&target) {
+                        State::Calling(Some(request), Some(service))
+                    } else {
+                        // Ensure that there is capacity for a new slot
+                        cache.poll_insert()?;
```

Comment:
What if this returns NotReady?

Reply:
It appears that poll_insert can't actually return `NotReady`... But the type system should actually reflect the possible outcomes (this is more important than the naming lining up, imo)...

I would suggest changing poll_insert to return a `Poll<(), Never>` and, instead, handling the not-ready->Error case in the ResponseFuture impl

Reply:
My [initial suggestion](https://github.com/linkerd/linkerd2-proxy/pull/247#discussion_r293576909) was that `poll_insert` should be renamed to `can_insert` (or something) and return `Result` rather than `Poll`, since (as written) it will never be `NotReady`. 

@kleimkuhler [pointed out](https://github.com/linkerd/linkerd2-proxy/pull/247) that @seanmonstar suggested _changing_ this function to be named `poll_insert` and return `Poll`, since it polls the timer internally and will thus fail if called outside of a task context, so the current approach is a bit of a compromise between our suggestions. 

IMO, I still think this should just return `Result` --- I'd rather not have the type signature include a case that will never actually be returned. But, I think that at this point, @olix0r, @seanmonstar, and I should probably try to come to an agreement on what the right thing to do here is, so that we can stop telling @kleimkuhler to change his code one way and then change it back...

Reply:
I think we are approaching the right decision. I feel it should return a `Poll` based off @seanmonstar comments, but I also like @olix0r's suggestion about a `Poll<(), Never>`. I have that change in there, and `NoCapacity` is constructed at the router now instead of the cache.

If `Result` does make sense after further discussion I can change it back! I don't mind the back and forth; good to see the reasons for each.